Added the IssuerPermissions back to solve an bug

roy-powerfinance opened the pull request:
Resolves the problem where issuers don't have access to send extrinsics to give self the permission to send extrinsics.
Added some unit tests back

roy-powerfinance committed:
Added the IssuerPermissions back to solve an bug
Added some unit tests back

cowboy-bebug reviewed (approved):
LGTM

roy-powerfinance committed:
initialise_issuer also grant the issuer that access

roy-powerfinance committed:
fixed a build problem

jordy25519 commented on prml/consortium-permission/src/mock.rs:
```diff
@@ -36,6 +36,33 @@ pub const ACCESS_TOPIC: &[u8; 6] = b"access";
+            );
+        }
+    }
+    /// Remove all self-claimed permissions from an issuer.
```
Not sure if comment wrong or implementation is wrong (maybe both)

/// Remove all self-claimed permissions from an issuer.
or
/// Remove self-claimed access permission for`issuer` (it does this afaict)

does it need the if block, since `ConsortiumPermission::claim((issuer, ACCESS_TOPIC.to_vec()));` would return `ACCESS_VALUE` false / unset if it doesn't exist

roy-powerfinance replied:
Roger. I'll update the comments and double check this

roy-powerfinance replied:
I think the if block is needed. Because the "access" can be granted by another person. We should only revoke "access" if it is granted by the "issuer" themselves when "access" is revoked

roy-powerfinance replied:
* Updated the comments here.

jordy25519 replied:
gotcha. I meant the outer if, as this function seemed made to remove self access claim but now I see it _could_ be called with any topic

roy-powerfinance replied:
I want to make it so that we only revoke the "access" permission if when we revoke their authority on the "access" topic.
Because we could be revoking their rights on other topics, such as mint_burn, and others, but still want them to keep their "access" right. Otherwise revoking any authority will revoke their "access" right, which seems very harsh
does that make sense?

jordy25519 reviewed (changes requested):
revoking part needs some reworking

jordy25519 commented on prml/consortium-permission/src/lib.rs:
```diff
@@ -203,6 +221,14 @@ decl_module! {
         pub fn force_remove_issuer(origin, who: T::AccountId) {
             ensure_root(origin)?;
+
+            // Notify the revocation of all current permissions.
```
this comment doesn't line up with the implementation of `revoke_issuer_permissions`

roy-powerfinance replied:
revoke_issuer_permissions can behave differently depending on its implementation. 
the module will simply notify the revocation by calling the revoke_issuer_permissions on all the topics it is revoking

jordy25519 replied:
makes sense!

roy-powerfinance committed:
Updated some comments

jordy25519 commented on prml/consortium-permission/src/mock.rs:
```diff
@@ -36,6 +36,34 @@ pub const ACCESS_TOPIC: &[u8; 6] = b"access";
+impl IssuerPermissions for IssuerPermissionsMock {
+    type AccountId = AccountId;
+    type Topic = Topic;
+    /// When an issuer is authorized to make claims on the "access" topic, also grant them the
+    /// "access" permission.
```
```suggestion
    /// When an issuer is authorized to make claims on the "access" topic, they also grant themselves
    /// "access" permission.
```

jordy25519 commented on prml/consortium-permission/src/mock.rs:
```diff
@@ -36,6 +36,34 @@ pub const ACCESS_TOPIC: &[u8; 6] = b"access";
+            );
+        }
+    }
+    /// When an issuer's authority on the "access" topic is revoked, also revoke their "access" permission.
+    fn revoke_issuer_permissions(issuer: &Self::AccountId, topic: &Topic) {
```
```suggestion
    /// When an issuer's authority on the "access" topic is revoked, also revoke their self-claimed "access" permission.
    fn revoke_issuer_permissions(issuer: &Self::AccountId, topic: &Topic) {
```

jordy25519 replied:
don't commit this not sure if whitespace is correct 😅

jordy25519 reviewed (commented):
looking good now
- mentioning self-claim is an important quirk imo
- missing one test for when `claim_issuer != issuer` (i.e. hit that if block)

roy-powerfinance committed:
Updated comments

roy-powerfinance committed:
Added a unit tests that tests the revocation of the topic permissions
Added a unit tests that tests force-remove and its revocation of
topic permissions

roy-powerfinance commented:
> looking good now
> 
> * mentioning self-claim is an important quirk imo
> * missing one test for when `claim_issuer != issuer` (i.e. hit that if block)

Yeah good point - I've fixed up the comments and added some more unit tests

jordy25519 reviewed (approved):
nice one
